### ISSUE TITLE
Android wallets - Insufficient energy when casting a vote

### ISSUE THREAD

**Issue by mh-concordium** [Concordium/concordium-governance-committee-voting#130]
Deploying locally all the components and trying to use the election dApp resulted with insufficient energy. I was trying to cast a vote with Concordium seedphrase wallet on Android 1.5.1 (70).

**To reproduce:**

1. Use the test script to deploy the election but comment out the voting function so that the script doesn't perform voting.
2. Run the election setup with min 5 nominees.
3. Build and deploy the dApp
4. Serve all the necessary files for the dApp.
5. Cast a vote with seedphrase wallet. Select candidate number 3.

**Result:**

Insufficient energy.

**Expected result:**

User should be able to cast a vote.

**Note:**

The issue is triggered if the number of nominees is 5. With 3 nominees, it is possible to cast a vote.

**To reproduce issue with Test-bench:**

1. Connect to [test bench](https://wallet-test-bench.testnet.concordium.com/) using wallet connect.
2. In the first case (**(IP) Testing simple input parameters**) select function: String.
3. In the input parameter paste the 60000 characters string that is in the file attached.
4. Select Set string value

**Result:**

Notice that at the time the signature page pops up in the wallet, there is a toast error message (communication error). The transaction fails, insufficient energy.

**Expected:**

Transaction passes successfully.

In this case energy assigned by the dApp should be used as the actual energy.

Attachments

[insufficient_energy.mp4](https://uploads.linear.app/430cc2b6-6b19-45f6-b7f0-6a0093e10c15/6e85626a-1562-4073-8f3f-487b92b6189a/1089e122-05b5-4100-9ba4-27d07813e0e6?signature=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJwYXRoIjoiLzQzMGNjMmI2LTZiMTktNDVmNi1iN2YwLTZhMDA5M2UxMGMxNS82ZTg1NjI2YS0xNTYyLTQwNzMtOGYzZi00ODdiOTJiNjE4OWEvMTA4OWUxMjItMDViNS00MTAwLTliYTQtMjdkMDc4MTNlMGU2IiwiaWF0IjoxNzEyOTEwODA3LCJleHAiOjMzMjgzNDcwODA3fQ.J90mrLWDhL5zdONjPg6R3XqjYSld37aYsa5KFIAEl98)

[large_string.txt](https://uploads.linear.app/430cc2b6-6b19-45f6-b7f0-6a0093e10c15/192196cd-afd9-4932-9980-9ed82e0f67ce/1256fff5-4324-4bb1-94ae-f0aaafe7155b?signature=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJwYXRoIjoiLzQzMGNjMmI2LTZiMTktNDVmNi1iN2YwLTZhMDA5M2UxMGMxNS8xOTIxOTZjZC1hZmQ5LTQ5MzItOTk4MC05ZWQ4MmUwZjY3Y2UvMTI1NmZmZjUtNDMyNC00YmIxLTk0YWUtZjBhYWFmZTcxNTViIiwiaWF0IjoxNzEyOTEwNzI4LCJleHAiOjMzMjgzNDcwNzI4fQ.1UOM3uqkaKjjODdZhylORR-uiUvogWV_pdtcsMwu2wY)

**Verify with deep linking as well.**

**Comment by Radiokot** [comment]
Is this the same issue we failed to reproduce in winter?

**Comment by abizjak** [comment]
As I understand the issue the problem comes from the following sequence of steps

When the request is received the wallet tries to "estimate transaction fee" here
```
                    proxyRepository.getTransferCost(
                        type = "update",
                        amount = payload.amount.toBigInteger(),
                        sender = walletConnectData.account!!.address,
                        contractIndex = payload.address.index,
                        contractSubindex = payload.address.subIndex,
                        receiveName = payload.receiveName,
                        parameter = payload.message,
                        success = {
                            walletConnectData.maxEnergy =
                                if (payload.maxEnergy != 0L) payload.maxEnergy else payload.maxContractExecutionEnergy
                            walletConnectData.cost = it.cost.toBigInteger()
                            walletConnectData.energy = it.energy
                            transactionFee.postValue(walletConnectData.cost)
                        },
                        failure = {
                            handleBackendError(it)
                        }
                    )
```

That fails when the URL is too large (when the parameter is too large).
Now the dapp is sending `maxContractExecutionEnergy` set to whatever value it determines is correct. So at this point `maxEnergy` is `null` as well as `cost` and `energy`.

Then in `createTransaction` we have
```

    private suspend fun createTransaction(keys: AccountData) {
        ...
        if (payload is Payload.ContractUpdateTransaction) {
            payload.maxEnergy = walletConnectData.energy ?: 0
            payload.maxContractExecutionEnergy = walletConnectData.maxEnergy ?: 0
        }
        ...
```
which means that both `maxEnergy` and `maxContractExecutionEnergy` are set to 0.

So basically the wallet should not be trying to do the estimation and it should
- if maxEnergy is set in the wallet connect payload use that and send it to the rust library call.
- if maxContractExecutionEnergy is set use that and send it to the rust library call

**Comment by abizjak** [comment]
@Radiokot why did you reopen this issue?

**Comment by Radiokot** [comment]
It is Linear integration 🙄

**Comment by mh-concordium** [comment]
I have reopened now to remember to retest it before we release the wallet.
